### ISSUE TITLE
Weight sharing between networks

### ISSUE THREAD

**Issue by pedronahum** [NervanaSystems/coach#59]
Hi,
After a quick study of your code, it is not clear to me how to create an agent with a complex network where the weights between networks are shared. 

Lets take as an example, UNREAL, https://deepmind.com/blog/reinforcement-learning-unsupervised-auxiliary-tasks/. 3 networks are built on top of a base A3C agent. 

Is currently supported? Or I just missed "reused flags" in the network wrapper. 

Thanks in advance

**Comment by itaicaspi-intel** [comment]
Hi Pedro,

Thanks for the question.
Take a look at the explanation of how the network structure is defined [here](http://coach.nervanasys.com/design/index.html#network-design).

In the case of the UNREAL agent, the convolution part and LSTM layer are shared, but there are several output heads for the network (pixel control, reward prediction, pi, etc.).

So, we will have a single network with a single observation embedder, an LSTM middleware and one output head for each auxiliary task, as well as a policy head and a V head for the A3C algorithm.

For example, the pixel control head (which will be defined [here](https://github.com/NervanaSystems/coach/blob/master/architectures/tensorflow_components/heads.py) ) will receive the LSTM output as its input, will have a deconvolution network and will have its own loss (L_PC).

The total loss for the network will be automatically computed as a sum of the losses of all the heads. The weighting terms (lambda_PC etc.) can be defined from the preset under agent.loss_weights.

Does it make sense?

**Comment by pedronahum** [comment]
Hi Itai,

Many thanks for getting back to me, greatly appreciated!

Your explanation makes perfect sense (particularly after spending some time reading the documentation). The challenge that I see is that some "complex" networks may need different inputs, yet they will need to share the weights. 

In the particular case of UNREAL, the input of the A3C network is the environment, yet the input for the other networks is a Replay Buffer. I will take a closer look at the examples (eg, Deep deterministic policy gradient) where I see that several observation embedders are created. 

Thanks again,

Pedro N.

**Comment by itaicaspi-intel** [comment]
If you want to train them simultaneously (in a single forward backward pass) then you are right. 
To train with a single pass, we might want to add a weights reuse mechanism to Coach. In that case you can use the `use_separate_networks_per_head` flag that will create a network were the input embedder and middleware are not shared between the heads. Reusing the weights between the two networks will enable doing the exact thing you described.

If you can compromise on training in two passes then there are two options:
* It may be possible to reuse the weights by creating two networks (one for A3C and one for the auxiliary tasks) with the same name. The tensorflow scoping we are using should take care of the rest and reuse the weights for the input embedder and middleware.
* Another option would be to use a single network but train it with two passes, where the first updates the A3C heads and the second updates the auxiliary tasks heads. The first pass, the network will be provided with an input from the environment, and the targets for the auxiliary heads will be their current value (such that the loss will be zeroed for them). For the second pass, the network will be provided with an input from the replay buffer, and the targets for the A3C heads will be their current value.

**Comment by pedronahum** [comment]
Hi, 
Greatly appreciate the detailed answer. I am sure I will not be the only one who will benefit from all the tips. 

I am closing the "issue" as Coach provides ways to address "weight sharing". Weather or not you would like to make it a "first class citizen" it is based on your priorities and next steps. 

Again, thanks for the information. Now it is my time to play with Coach. 

Best
